File: src/compiler/syntax/statement_ast.rs
use crate::compiler::lexer::span::Span;

use super::{ast_node::ASTNode, expression_ast::ExpressionAST};

#[derive(Debug)]
pub enum StatementAST {
    Print {
        expression: Box<ExpressionAST>,
        span: Span,
    },

    Expression {
        expression: Box<ExpressionAST>,
        span: Span,
    },
    If {
        condition: Box<ExpressionAST>,
        then_branch: Box<StatementAST>,
        else_branch: Option<Box<StatementAST>>,
        span: Span,
    },
    WhileLoop {
        condition: Box<ExpressionAST>,
        block: Box<StatementAST>,
        span: Span,
    },
    Block {
        declarations: Vec<ASTNode>,
        span: Span,
    },
    Break(Span),
    Continue(Span),
}

File: legacy-java/src/main/java/com/kaori/compiler/syntax/StatementAST.java
package com.kaori.compiler.syntax;

import java.util.ArrayList;
import java.util.List;

public interface StatementAST extends DeclarationAST {
	record Print(int line, ExpressionAST expression) implements StatementAST {
	}

	record Expr(int line, ExpressionAST expression) implements StatementAST {
	}

	record Block(int line, List<DeclarationAST> declarations) implements StatementAST {
		public Block(int line) {
			this(line, new ArrayList<>());
		}
	}

	record If(
			int line,
			ExpressionAST condition,
			Block thenBranch,
			StatementAST elseBranch) implements StatementAST {
	}

	record WhileLoop(
			int line,
			ExpressionAST condition,
			Block block) implements StatementAST {
	}
}
